src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Attr.java

```diff
@@ -1724,14 +1730,22 @@ private void handleSwitch(JCTree switchTree,
                                 rs.basicLogResolveHelper = prevResolveHelper;
                             }
                         } else {
-                            Type pattype = attribExpr(expr, switchEnv, seltype);
+                            ResultInfo valTypInfo = new ResultInfo(KindSelector.VAL_TYP,
+                                                                   !seltype.hasTag(ERROR) ? seltype
+                                                                                          : Type.noType);
+                            Type pattype = attribTree(expr, switchEnv, valTypInfo);
                             if (!pattype.hasTag(ERROR)) {
-                                if (!stringSwitch && !types.isAssignable(seltype, syms.intType)) {
-                                    log.error(pat.pos(), Errors.ConstantLabelNotCompatible(pattype, seltype));
-                                }
                                 if (pattype.constValue() == null) {
-                                    log.error(expr.pos(),
-                                              (stringSwitch ? Errors.StringConstReq : Errors.ConstExprReq));
+                                    Symbol s = TreeInfo.symbol(expr);
+                                    if (s != null && s.kind == TYP && allowPatternSwitch) {
+                                        log.error(expr.pos(),
+                                                  Errors.PatternExpected);
+                                    } else {
+                                        log.error(expr.pos(),
+                                                  (stringSwitch ? Errors.StringConstReq : Errors.ConstExprReq));
+                                    }
+                                } else if (!stringSwitch && !types.isAssignable(seltype, syms.intType)) {
+                                    log.error(pat.pos(), Errors.ConstantLabelNotCompatible(pattype, seltype));
```

Comment:
I wonder if in the long term it's not better instead of having several ad-hoc way to find the kind of switch it is,
`types.isAssignable(seltype, syms.intType)`, `stringSwitch` or `pattype.constValue() == null && s.kind == TYP`
to have an enum with 4 values INT_SWITCH, STRING_SWITCH, ENUM_SWITCH and PATTERN_SWITCH.

Reply:
I apologize for a belated answer. It is my plan to try to categorize switches, but when considering error recovery and various desugaring stages, it is not trivial. Also, please note that some of the conditions are not a property of the switch, but rather a property of the case (like in this case - saying `case String:` instead of `case String s:` is more a property of the case, than the switch.

Reply:
Yes, it's not trivial, and it's not for that PR, but it's usually a good idea in a compiler to have such kind of canonical representation because it avoid bugs between different passes that does those checks in various non compatible ways. Especially when you have different desugaring stages.